coffee/rewriter.py

```diff
@@ -607,60 +708,63 @@ def _hoist(self, expansion, exp, grp):
                                                    'i': len(self.expanded_decls)}
         decl = dcopy(hoisted_decl)
         decl.sym.symbol = hoisted_exp.symbol
+        stmt = Assign(hoisted_exp, expr)
         # Update the AST
-        hoisted_loop.body.append(Assign(hoisted_exp, expr))
+        hoisted_loop.body.append(stmt)
         insert_at_elem(hoisted_place.children, hoisted_decl, decl)
         # Update tracked information
         self.expanded_decls[decl.sym.symbol] = decl
-        self.hoisted[hoisted_exp.symbol] = (expr, decl, hoisted_loop, hoisted_place)
-        self.expr_graph.add_dependency(hoisted_exp, expr, 0)
+        self.hoisted[hoisted_exp.symbol] = (stmt, decl, hoisted_loop, hoisted_place)
+        self.expr_graph.add_dependency(hoisted_exp, expr)
+        self.cache[cache_key] = hoisted_exp
         return {exp: hoisted_exp}
 
     def _expand(self, node, parent):
```

Comment:
Yes, but I wrote it before Lawrence wrote the infrastructure, so maybe one day we can turn it into a visitor. Not gonna happen now.